redmine.py

```diff
@@ -1,74 +1,106 @@
+    Find the "subject" string in the JSON data of a ticket.
+    :param ticket_url: URL for a ticket
+    :param api_key: API secret key (string), or None if you do not want to
+                    authenticate to Redmine.
+    :returns: twisted.internet.defere.Deferred. When this Deferred fires, it
+              will return a "subject" string to its callback.
     """
-    if not match:
-        return ''
-    ticket_id = match[-1]  # Always the last one in the group
-    ticket_id = ticket_id.strip()  # probably not necessary?
-    return ticket_id.strip('#')
+    api_url = "%s.json" % ticket_url
+    request_headers = {}
+    if api_key is not None:
+        request_headers['X-Redmine-API-Key'] = api_key
+    try:
+        response = yield treq.get(api_url, headers=request_headers, timeout=5)
+        if response.code != 200:
+            subject = 'could not read subject, HTTP code %i' % response.code
+            defer.returnValue((ticket_url, subject))
+        else:
+            content = yield treq.json_content(response)
+            defer.returnValue((ticket_url, content['issue']['subject']))
+    except Exception, e:
+        # For example, if treq.get() timed out, or if treq.json_content() could
+        # not parse the JSON, etc.
+        subject = 'could not read subject, %s' % e.message
+        defer.returnValue((ticket_url, subject))
 
-def get_ticket_response(api_url, api_key):
-    if api_key:
-        request_headers = {'X-Redmine-API-Key': api_key}
+def construct_message(urls_and_subjects, nick):
+    """
+    Return a string about a nick and a list of tickets' URLs and subjects.
+    """
+    msgs = []
+    for url_and_subject in urls_and_subjects:
+        ticket_url, subject = url_and_subject
+        msgs.append('%s [%s]' % (ticket_url, subject))
+    if len(msgs) == 1:
+        msg = msgs[0]
     else:
-        request_headers = {}
-    return requests.get(api_url, headers=request_headers)
+        msg = "{} and {}".format(", ".join(msgs[:-1]), msgs[-1])
+    return "%s might be talking about %s" % (nick, msg)
 
+def send_message(urls_and_subjects, client, channel, nick):
+    """
+    Send a message to an IRC/XMPP channel about a list of tickets' URLs and
+    subjects.
+    """
+    msg = construct_message(urls_and_subjects, nick)
+    client.msg(channel, msg)
 
-def get_issue_subject(response):
-    """Find the "subject" string in the JSON data of python-requests' response
-    object."""
-    try:
-        result = response.json()
-    except ValueError as err:
-        result = {}
-        logger.error("couldn't access that URL. response was %s: %s" % (response.status_code, err))
-
-    try:
-        return result['issue']['subject']
-    except KeyError:
-        return 'unable to read subject'
+def match_tickets(message):
+    tickets = []
+    pattern = re.compile(r"""
+       (?:               # Prefix to trigger the plugin:
+            issues?      #   "issue" or "issues"
```

Comment:
The `?` just makes the preceding token optional. So it's saying "make `s` optional" there.